### DISCUSSION TITLE
Tutorial code blocks, mqtt, is not "good practice"

### DISCUSSION THREAD

**Discussion by sfromis** [tasmota/Berry_playground#28] [Tutorials and Code Blocks]
This line from the code block is not a good idea:
`tasmota.add_rule("MQTT#Connected=1", subscribes)`

This is not really useful, as `mqtt.subscribe` can since quite some time be done before Tasmota has connected to MQTT, as Tasmota will automatically manage the calls to the broker at the right time.

Even before this enhancement, the example would be problematic, as it could lead to multiple subscriptions in case of Tasmota reconnecting without restarting, like in case of unstable Wifi. The difficulties of doing a robust subscription management was the impetus for this being integrated in Tasmota Berry, instead of users expected to do the relevant subscribe/unsubscribe at good points in time.

Minor detail, the example is doing `import string` without this being used...

**Comment by darrylb123** [comment]
Ah yes, the danger of looking at code snippets on the discussion forum.
I'll see if I can get it to work without it.

**Comment by wjohn007** [comment]
> This is not really useful, as `mqtt.subscribe` can since quite some time be done before Tasmota has connected to MQTT, as Tasmota will automatically manage the calls to the broker at the right time.
> 

@sfromis  
Thank you for clarification.

This information is very important for users.
How can a normal user find such information in the documentation?

What are the negative effects of a redundant "import" ?

John

**Reply by wjohn007** [reply]
I found all information about berry's mqtt skills in  [mqtt.be](https://github.com/arendst/Tasmota/blob/0ac5d0945b6d1d8a1cacb5bf3f9d5cd360180093/lib/libesp32/berry_tasmota/src/embedded/mqtt.be)

John
